.git-hooks/pre-commit

```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+# Setup the pre-commit symlink as `ln -s ../../.git-hooks/pre-commit .git/hooks/pre-commit`
+
+while :
+do
+  echo "Running pre-commit hook"
+  # Run the pre-commit hook
+  pre-commit run -a
+
+  # Check if there are any changes
+  if git diff --quiet; then
+    echo "No changes after running pre-commit."
+    break  # Exit the loop if there are no changes
+  else
+    # Stage the changes
+    git add .
```

Comment:
Sure! You're correct, but I'm still unsure about the way to input files. Pre-commit checks automatically reformat all the files in the repository and fixes them. This makes it tricky to specify which files to include for a couple of reasons:
1. There are a lot of files that get reformatted, and it's hard to know in advance which ones.
2. At the start, we can't predict which files will need formatting.

I've thought of three possible solutions:
1. We could ask to specify which formatted files we want to include when the pre-commit is in execution.
2. The pre-commit output can tell us which files need formatting and will be fixed, so we could make the script use that information to include only those files (Unsure about how to actually include the required part from a bunch of fancy output)
3. Alternatively, we could let input a list of files we don't want to include in the pre-commit.

Reply:
Why do we need to do this? Why can't I just do a git add on whatever files I want to commit?

Reply:
There's a way we can only stage modified files `git add -u` and not untracked ones

Reply:
@ErnestoLoma Pre-commit does auto-formatting and with every change we need to `git add` the modified files by pre-commit since it fails the underlying formatters For example black or prettier automatically reformat code and then pre-commit hook fails after detecting changed files. After that I have to manually stage changed files and run commit once again, unnecessary friction since formatting just works and changes are expected.
Does this makes sense?

Reply:
@0xquark Like I believe @cmgrote said this could result in files being committed that I don't want to commit. I prefer the small bit of friction to loss of control.

Reply:
I understand that would be a big problem! I found this `git diff --name-only --cached | xargs git add` as a solution to the problem to only commit files that are modified by the pre-commit hook!
This problem happens with a lot of people using pre-commit since pre-commit doesn't allow modifying things in staging area.

Reply:
I am okay with keeping things the way they are because I agree that automatically committing changes without a person checking them can be risky. I brought up the idea because it can be tiresome to manually stage changes on  after running the pre-commit  and then running it again but also things are pretty much checked at CI level too so there's that for safety.
Ref: https://github.com/pre-commit/pre-commit/issues/747#issuecomment-386782080

Reply:
So trying to summarize, would it work like this?

1. We decide which files to stage for a commit using `git add ...` (as normal)
2. We then run the `git commit -s -m '...'`
3. The script / pre-commit hook then executes the various linters.
4. The script / pre-commit hook then identifies _only any files that the pre-commit hook itself modifies_ (in 3).
5. The script / pre-commit hook then re-stages _only those files it changed_ (identified in 4).
6. The script / pre-commit hook then tries to commit (again) those re-staged files.

^^ If it can work like that, then I think we have the best of both worlds?

Reply:
Sounds good to me — curious to give it a go, perhaps, before we merge it into main, though...

@ErnestoLoma any other gotchas you can think of given the flow outlined above?

(It's a small bit of friction, but adds up — particularly every time we re-generate since there are loads of files involved — so I'm very eager to remove that friction if we can.)

Reply:
Here's a real-life example: I made changes to two files, one called `pre-commit` and another called `check_tag.py.` However, when I used the `git add` command, I only added the "pre-commit" file to the staging area. Then, when I tried to commit these changes and the pre-commit hook ran, it only ran on the "pre-commit" file and didn't do anything to "check_tag.py." As a result, only the "pre-commit" file got committed, and "check_tag.py" remained unchanged.

<img width="947" alt="image" src="https://github.com/atlanhq/atlan-python/assets/84588720/3ab7ed18-6145-4f40-9d81-18948b43bf24">

` But i still want someone to cross check this, if this works really as expected and is not a mere coincidence`